utils/build_rpms_from_scratch.sh

```diff
@@ -0,0 +1,261 @@
+  fi
+
+  # Ignore dpdk because want to use RPMs built from daos-stack/dpdk
+  zypper update -y --skip-interactive --no-recommends
+}
+
+prepare_env_rocky() {
+  el_version=$OS_VERSION # default to '8'
+
+    # Install needed packages
+    # git so can clone the repo
+    # rpm-build, rpmdevtools, createrepo for building the rpms and setting up the rpm repo
+    # dnf-plugins-core for additional dnf capabilities
+    # make is used for managing the rpmbuild process. This should probably be a dependency of the rpm.spec files, but not
+  dnf install -y \
+      git \
+      rpm-build \
+      rpmdevtools \
+      createrepo \
+      dnf-plugins-core \
+      epel-release \
+      make # shouldn't be needed here, but dependency spec files don't include
+
+  # enable Power Tools or CodeReady Builder depending on el version. Needed for yasm and maybe others
+  if [[ "$el_version" == "8" ]]; then
+    dnf config-manager --set-enabled powertools
+  elif [[ "$el_version" == "9" ]]; then
+    dnf config-manager --set-enabled crb # CodeReady Builder
+  fi
+
+  # ignore dpdk because want to use rpms build from daos-stack/dpdk
+  dnf update -y --exclude=dpdk --exclude=dpdk-devel
+}
+
+l_createrepo() {
+  createrepo "$1"
+
+  if [[ "$OS_ID" == "opensuse-leap" ]]; then
+    zypper refresh
+  fi
+}
+
+prepare_env() {
+  if [ -f /tmp/setup_done ]; then return; fi
+
+  rpms_dst=$1
+  build_dst=$2
+
+  mkdir -p "$rpms_dst" "$build_dst"
+
+  if [[ "$OS_ID" == "opensuse-leap" ]]; then
+      prepare_env_leap
+    else
+      prepare_env_rocky
+  fi
+
+  setup_local_repo "$rpms_dst"
+  l_createrepo "$rpms_dst"
+
+  touch /tmp/setup_done
+}
+
+verify_os_version() {
+  $OS_ID
+  if [[ "$OS_ID" == "opensuse-leap" ]]; then
+    if ! [[ "$OS_VERSION" == "15.4" || "$OS_VERSION" == "15.5" || "$OS_VERSION" == "15.6" ]]; then
+      echo "Untested SUSE Version: $suse_version"
+      exit 1
+    fi
+  elif [[ "$OS_ID" == "rocky" ]]; then
+    if ! [[ "$OS_VERSION" == "8.9" || "$OS_VERSION" == "9.3" ]]; then
+      echo "Untested Rocky Version: $suse_version"
+      exit 1
+    fi
+    else
+      echo "Untested OS: $OS_ID"
+      exit 1
+  fi
+}
+
+# -----------------------------------------------------------------------------
+# The following functions should be common, and not have OS specific calls
+# -----------------------------------------------------------------------------
+
+# Function to clone a daos-stack repo, build rpms, and setup the rpm repository
+build_and_copy_rpm() {
+  local repo_name=$1
+
+  # Check if already successful - helps if running script many times while troubleshooting
+  if [ ! -f "$repo_name"/success ]; then
+    if [ -d "$repo_name" ]; then rm -rf "$repo_name"; fi
+    echo "Building $repo_name ..."
+    git clone https://github.com/daos-stack/"$repo_name".git
+    cd "$repo_name"
+      install_dependencies "$repo_name".spec
+      make rpms RPM_BUILD_OPTIONS="--nocheck" # --nocheck = don't run tests during rpm build process
```

Comment:
```suggestion
      MAKEARGS=()
      if [ -f Makefile-rpm.mk ]; then
        MAKEARGS=(-f Makefile-rpm.mk)
      fi
      if [ -d utils/rpms ]; then
        MAKEARGS+=(-C utils/rpms)
      fi
      make "${MAKEARGS[@]}" rpms RPM_BUILD_OPTIONS="--nocheck" # --nocheck = don't run tests during rpm build process
```
will let you use this for _raft_ and _daos_ also and be future-proof against any other project that utilizes these exceptions.

Reply:
I think it's a little more complicated than this. build_and_copy_rpm also clones the repo and because raft is a submodule of daos, the daos repo must be cloned first, then raft rpms built, then daos rpms can be built. So a bit of a chicken and egg issue with the way I wrote build_and_copy_rpms. I appreciate the help trying to remove the redundancy, though. I've probably already spent too much time trying to figure out how to remove the duplicate code without making things too complicated.

Reply:
> I think it's a little more complicated than this.

I don't think it is.

> build_and_copy_rpm also clones the repo and because raft is a submodule of daos, the daos repo must be cloned first, then raft rpms built, then daos rpms can be built.

You can clone and build the raft repo independently of daos just like you do the other dozen repos.  There is no need to clone daos and build raft out of the daos clone.

Reply:
My concern is getting the correct raft commit that's actually used for daos. The 'fun' of it being a submodule is it doesn't have to be the tip of master. Right now it looks like the submodule daos uses is 1 commit behind the tip of master.

Reply:
Yes, this is bad and I have said so any number of times to the developers of raft.  Our dependency policy is that we always use the tip of master of all of our dependencies.  That is how our users know what to use while we do not unnecessarily (IMO) complicate our dependency handling with tags and branches.

Commits in any dependency that are not ready for use should be in branches/PRs, just like we do for every other project (esp. daos).  We have all of the capabilities that we need to do any kind of development or testing in CI from branches/PRs so there is no reason to be pushing things to a project's master that we don't intend to be what a production (or current development) site to use.

We need to stop allowing raft to be treated exceptionally in this process/policy as it only serves to confuse and make us have to have exceptions to policies with no good reason.